### ISSUE TITLE
New language support: Nim

### ISSUE THREAD

**Issue by sealmove** [kaitai-io/kaitai_struct#619]
Support for the Nim programming language is now under development.
Here is a first draft of the [runtime library](https://github.com/kaitai-io/kaitai_struct_nim_runtime).

**Comment by deem0n** [comment]
Is it ready for general use?

**Comment by sealmove** [comment]
> Is it ready for general use?

Not yet, but I am actively working on it. Glad for your interest!

**Comment by deem0n** [comment]
I planned to use C++ runtime and hand-made nim wrapper for it, but native nim runtime would be much better!

**Comment by sealmove** [comment]
Yeah, I am trying to use as many native Nim features as possible, and produce as much human readable code as possible, so it takes some time.

**Comment by sealmove** [comment]
Though I can't yet estimate a time frame for when it will be usable, so you probably shouldn't wait this time, especially if those wrappers are easy to make.

**Comment by GreyCat** [comment]
#631 — a discussion on grand rework of Nim backend.

**Comment by GreyCat** [comment]
I see that nim appeared in [CI dashboard](https://ci.kaitai.io/) — although currently it is at 0%, it's a great step forward! Please tell me if I can be of any help with infrastructure?

**Comment by GreyCat** [comment]
I've noticed that Nim test run is currently failing due to compilation errors. Any chance that [partial builder](https://github.com/kaitai-io/kaitai_struct_tests/tree/master/builder) framework would of of any help here — i.e. doing the routine "create project" => "try to compile" => "exclude unbuildable files" => "create project again" => ad inf?

**Comment by sealmove** [comment]
@GreyCat thanks!

I am not 100% sure how partial builder works, but I suspect it can't be used for Nim at the moment because the way I've set up testing hackish. Let me explain:

For testing I am using [unittest module](https://nim-lang.org/docs/unittest.html). This module provides minimal features, and thus is not super convenient, but I decided to go with it because it is part of the standard library.

**Motivation for the hack:**
The module produces one report for each unittest file. This is unfortunate because it can't produce (at least to my knowledge) a single report based on all the files in `kaitai_struct/tests/spec/nim/tests`.

**The hack:**
[`generate_test_suite`](https://github.com/kaitai-io/kaitai_struct_tests/blob/master/spec/nim/generate_test_suite.nim) collects code snippets from `kaitai_struct/tests/spec/nim/tests` and assembles one big unittest file by adding:
  - imports
  - helper procs for looser comparisons where needed
  - the unittest macros for producing `$TEST_OUT_DIR/nim/report.xml` 

**Why I think partial builder can't work at the moment**:
The files in `kaitai_struct/tests/spec/nim/tests` are not directly compilable, but only minimal code snippets which consist of:
  - a statement to obtain the object
  - the checks

In conclusion: the step "exclude unbuildable files" can't be easily applied.

**2 kinds of errors**
When running tests this way, there are 2 kinds of compilation errors:
  1. Some of the files in `"$FORMATS_COMPILED_DIR"/nim` fail to compile
  2. The test suite itself fails to compile

**Possible solutions:**
  1. Re-implement the above functionality in partial builder and augment it so that it can produce "the perfect" unittest file.
  2. Simply augment generate_test_suite with a partial builder mechanism (easier but eh... ugly).

I think solution 1 is much neater because in a sense it hides the above Kaitai-Struct-specific hack behind a Kaitai-Struct-specific mechanism, getting rid of ridiculously custom programs like generate_test_suite. The only issue is I've never written a single line of code in Ruby, but oh well, how hard could it be?

**Comment by cugu** [comment]
@sealmove Sorry for interruption here but please add a license to the runtime.

**Comment by sealmove** [comment]
It's fixed with a rework of the testing mechanism 410eba32812d2b59984c0850f15c902fa23dd42d

**Comment by sealmove** [comment]
> @sealmove Sorry for interruption here but please add a license to the runtime.

Done.

**Comment by sealmove** [comment]
Due to Nim's extensive metaprogramming capabilities, I realized a lib implementation is more suitable. Nim has the potential to eliminate the need of an external .ksy -> .nim compiler. KS-yaml code can be written directly in .nim files and a **library** can transform it into a parser **in memory**. This has several advantages:
- Safer implementation since it's done on AST level instead of source code.
- Better code integration

**Comment by sealmove** [comment]
Development focus has been shifted towards [nimitai](https://github.com/sealmove/nimitai).
Unfortunately this means Nim support for standard Kaitai Struct project will be delayed.

**Comment by GreyCat** [comment]
@sealmove, thanks for your update! Can I ask you to add it to [awesome-kaitai list](https://github.com/kaitai-io/awesome-kaitai)?

**Comment by sealmove** [comment]
> @sealmove, thanks for your update! Can I ask you to add it to [awesome-kaitai list](https://github.com/kaitai-io/awesome-kaitai)?

Sure, I will do that!

**Comment by KOLANICH** [comment]
What are the plans on supporting nim runtime in the compiler? Should I add explicit support of nim-related arguments to `kaitaiStructCompile.py`?

**Comment by sealmove** [comment]
I don't see a reason to remove it. It might take quite a while, but at some point in time -after nimitai is finished- and having accumulated knowledge from nimitai, it will be very easy for me to finish it for the KSC too.

**Comment by KOLANICH** [comment]
> How is kaitaiStructCompile.py related?

It is a python lib wrapping calls to KSC (2 backends are implemented, CLI and JVM, but JVM one is not distributed (#466) though it is more secure and much faster because KSC is GPLed and I don't want to take part in GPL proliferation since GPL proliferation is copyright proliferation) and fetching. It is usually used to compile specs into python files when building pip packages (it has a setuptools plugin for it), but it is possible to use it to compile specs into another languages, for example as a part of build system (i.e. to write a plugin for meson) or when building a python package requires to build some code other than python (i.e. cexts implemented in C++ or rust).

**Comment by sealmove** [comment]
After 4 months of stupid ambition, I realized [nimitai](https://github.com/sealmove/nimitai) has failed. The idea was to re-implement Kaitai Struct as a Nim library. As it turns out, there is no compile-time yaml parser for Nim (or any other language actually, afaik). Implementing one + re-implementing Kaitai Struct with Nim macros turned out to be way too much for its benefits. Therefore I am coming back to the original plan!

I implemented a tiny tool called [testify](https://github.com/sealmove/testify) which takes care of the testing infrastructure (it was easier for me than figuring out how to adjust `partial_builder`). The results are already visible on [dashboard](https://ci.kaitai.io/) (yeah I know, 3% is bad), so keep an eye!

**Comment by sealmove** [comment]
Ok, more than 60% of the tests pass. There are still a lot of things to adjust, but most features have been implemented. If you don't mind making some adjustments to the generated code, then you can already start using the back-end! (@deem0n). Soon I will write the page with Nim-specific notes explaining the API in detail. The API is different from the other back-ends' since Nim does not encourage object-oriented programming and I didn't want to make it feel _foreign_ to Nim users. I think you will find it intuitive, direct, and easy to use.
